CVLByExample/ExtensionContracts/README.md

```diff
@@ -0,0 +1,54 @@
+# Extension Contracts
+
+A contract extension is about splitting the functionality of a single contract into several ones and using delegatecalls between them.
+
+For example:
+```
+contract Extended {
+    address public extensionAddress; // Address of contract Extension
+
+    // Fallback function will handle all calls that don't match a function signature
+    fallback() external payable {
+        // Delegate the call to contract Extension
+        (bool success, ) = bAddress.delegatecall(msg.data);
+        require(success, "Delegatecall failed");
+    }
+}
+
+Contract B {
+	function foo() external { ... }
+}
+```
+Then when someone calls B(aAddress).foo() the call is delegated to B via A's fallback function.
+
+Usually the prover doesn't "understand" that B is in an extension of A, so the call above is unresolved.
+To support this pattern a new option `contract_extensions` was added which accepts a dictionary of the form:
+```
+"contract_extensions": {
+    "A": [
+        {
+            "extension": "B",
+            "exclude": ["value", "sender", "num"],
+        },
+        {
+            "extension": "C",
+            "exclude": ["value", "sender", "num"],
+        },
+    ],
+	"D": [
+		...
+	],
+    ...
+}
+```
+The keys to the dictionary are the contracts we want to extend, and their values are lists of extension (extension) contracts and lists of functions that should be excluded from the extension (to avoid duplicate functions).
+
+What this does is "transfer" the methods from the extension contracts into the extended one, so analysis could do its thing.
+
+Some more details:
+* The implementation of all external methods are removed from the extension contract, so they should never be called. To prevent this there are several guards:
+    - Parametric rules will filter out functions from the extension function directly (only as functions of the extended contract).
+    - Attempting to alias an Extension function will lead to a typechecker error.
+        + One shouldn’t call the functions of the extension contract using that contract as the receiver, only use the extended contract as a receiver.
+    - Attempting to add a preserved block on an extension function will lead to a typechecker error.
+        + Again, one should be using the extended contract as the receiver.
```

Comment:
all of this is very confusing, can we give examples: 
what is allowed? 
what is not allowed and what is the error message? 
So if there is no specific function in the top level that wraps a specific delegate call

Reply:
Honestly I'm not sure. I haven't done a super deep dive into how this feature works, I took this explanation from what Naftali wrote on his PR and I patched it a little.
Since we're in the middle of the monitoring project but we still want to have some example this feature before the release I think the right approach is to delete this section from the readme and open new tasks for extending this example with more examples and information if we'll really need to.
@yoav-el-certora what do you think?

Reply:
@naftali-g This information is also missing in the [Documentation PR](https://github.com/Certora/Documentation/pull/300/files)

Can you please assist?

Reply:
> So if there is no specific function in the top level that wraps a specific delegate call

Can you please elaborate?

Reply:
but in the methods block it is the extension? 

Liav, it should be nice if you can add a 'use built-in rule sanity' and show how it looks in the call trace?  is it all under the base contract?

Reply:
Yes it's all under the base contracts: https://vaas-stg.certora.com/output/15800/8de64e410516472cbc8f7d6317059680?anonymousKey=b4b480e8209e59136b4daa281aa5159a1fc66fd1